combinator/src/main/java/com/iluwatar/combinator/StudentValidation.java

```diff
@@ -0,0 +1,125 @@
+/**
+ * The MIT License
+ * Copyright (c) 2014-2016 Ilkka Seppälä
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.iluwatar.combinator;
+
+import java.text.DateFormat;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+final class ValidationSupport {
+    private static final ValidationResult valid = new ValidationResult(){
+        public boolean isValid(){ return true; }
+        public Optional<String> getReason(){ return Optional.empty(); }
+    };
+
+    static ValidationResult valid(){
+        return valid;
+    }
+}
+
+interface ValidationResult{
+    static ValidationResult valid(){
+        return ValidationSupport.valid();
+    }
+
+    static ValidationResult invalid(String reason){
+        return new Invalid(reason);
+    }
+
+    boolean isValid();
+
+    Optional<String> getReason();
+}
+
+final class Invalid implements ValidationResult {
+
+    private final String reason;
+
+    Invalid(String reason){
+        this.reason = reason;
+    }
+
+    public boolean isValid(){
+        return false;
+    }
+
+    public Optional<String> getReason(){
+        return Optional.of(reason);
+    }
+}
+
```

Comment:
Discussion: One downside from what I can see is that for each `String` in my bean such as Name, Surname, etc I need to create separate validation methods like `nameIsNotEmpty()`, `surnameIsNotEmpty()` this is doable for small beans but will require a lot of work for larger beans.
Can we have a generic `isNotEmpty(...)` validation that works on `Supplier`

A sample:
`static StudentValidation isNotEmpty(Supplier<String> supplier, String message) {
    return holds(student -> !supplier.get().trim().isEmpty(), message);
}`

Client code:
`StudentValidation.isNotEmpty(student::getName, "Name is empty")
    .and(StudentValidation.isNotEmpty(student::getSurname, "Surname is Empty")
`

The downside of this approach is message needs to be passed from outside.

Note I have not tried the code in IDE, so pardon me for any syntactical issues.